File: secret-santa-iac/function/pairing.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"

	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/s3"
)

const (
	bucketName = "secret-santa-pairings-dev" // Change based on your stage
)

type TokenData struct {
	Token  string `json:"token"`
	Gifter string `json:"gifter"`
	Giftee string `json:"giftee"`
	Status string `json:"status"`
}

func handler(request events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
	token := request.QueryStringParameters["token"]
	if token == "" {
		return events.APIGatewayProxyResponse{
			StatusCode: 400,
			Body:       `{"error": "Token is required"}`,
		}, nil
	}

	sess := session.Must(session.NewSession())
	s3Client := s3.New(sess)

	tokenData, err := getTokenData(s3Client, token)
	if err != nil {
		if err != nil && err.Error() == s3.ErrCodeNoSuchKey {
			return events.APIGatewayProxyResponse{
				StatusCode: 404,
				Body:       `{"error": "Token not found"}`,
			}, nil
		}
		log.Println("Error retrieving token:", err)
		return events.APIGatewayProxyResponse{StatusCode: 500, Body: `{"error": "Internal server error"}`}, nil
	}

	if tokenData.Status == "used" {
		return events.APIGatewayProxyResponse{
			StatusCode: 403,
			Body:       `{"error": "Token already used"}`,
		}, nil
	}

	tokenData.Status = "used"
	err = updateTokenData(s3Client, token, tokenData)
	if err != nil {
		log.Println("Error updating token status:", err)
		return events.APIGatewayProxyResponse{StatusCode: 500, Body: `{"error": "Failed to update token status"}`}, nil
	}

	responseBody, _ := json.Marshal(map[string]string{
		"gifter": tokenData.Gifter,
		"giftee": tokenData.Giftee,
	})

	return events.APIGatewayProxyResponse{
		StatusCode: 200,
		Body:       string(responseBody),
	}, nil
}

func getTokenData(s3Client *s3.S3, token string) (*TokenData, error) {
	key := fmt.Sprintf("%s.json", token)
	result, err := s3Client.GetObject(&s3.GetObjectInput{
		Bucket: aws.String(bucketName),
		Key:    aws.String(key),
	})
	if err != nil {
		return nil, err
	}
	defer result.Body.Close()

	var tokenData TokenData
	err = json.NewDecoder(result.Body).Decode(&tokenData)
	if err != nil {
		return nil, err
	}

	return &tokenData, nil
}

func updateTokenData(s3Client *s3.S3, token string, tokenData *TokenData) error {
	key := fmt.Sprintf("%s.json", token)
	body, err := json.Marshal(tokenData)
	if err != nil {
		return err
	}

	_, err = s3Client.PutObject(&s3.PutObjectInput{
		Bucket: aws.String(bucketName),
		Key:    aws.String(key),
		Body:   aws.ReadSeekCloser(bytes.NewReader(body)),
	})
	return err
}

func main() {
	lambda.Start(handler)
}

File: secret-santa-iac/pairing.py
import json
import boto3
import logging
import uuid
from botocore.exceptions import ClientError

# Set up logging
logger = logging.getLogger()
logger.setLevel(logging.INFO)

s3_client = boto3.client('s3')
BUCKET_NAME = 'secret-santa-pairings-dev'  # Adjust based on your environment

def handler(event, context):
    http_method = event['httpMethod']
    
    if http_method == 'POST':
        return create_token(event)
    elif http_method == 'GET':
        return get_pairing(event)
    else:
        return {
            "statusCode": 405,
            "headers": {
                "Access-Control-Allow-Origin": "*"
            },
            "body": json.dumps({"error": "Method not allowed"})
        }

def create_token(event):
    body = json.loads(event['body'])
    gifter = body.get("gifter")
    giftee = body.get("giftee")
    token = str(uuid.uuid4())  # Generate a unique token

    token_data = {
        "token": token,
        "gifter": gifter,
        "giftee": giftee,
        "status": "unused"
    }

    try:
        s3_client.put_object(
            Bucket=BUCKET_NAME,
            Key=f"{token}.json",
            Body=json.dumps(token_data),
            ContentType="application/json"
        )
    except ClientError as e:
        logger.error(f"Error creating the token file in S3: {error_code}")
        return {
            "statusCode": 500,
            "headers": {"Access-Control-Allow-Origin": "*"},
            "body": json.dumps({"error": "Failed to create token", "details": str(e)})
        }

    return {
        "statusCode": 200,
        "headers": {
            "Access-Control-Allow-Origin": "*"
        },
        "body": json.dumps({"token": token})
    }

def get_pairing(event):
    token = event.get("queryStringParameters", {}).get("token")
    if not token:
        logger.error("No token provided in request.")
        return {"statusCode": 400, 
                "headers": {"Access-Control-Allow-Origin": "*"},
                "body": json.dumps({"error": "Token is required as query param"})
                }

    try:
        response = s3_client.get_object(Bucket=BUCKET_NAME, Key=f"{token}.json")
        token_data = json.loads(response['Body'].read())
        
        if token_data["status"] == "used":
            return {
                "statusCode": 403, 
                "headers": {"Access-Control-Allow-Origin": "*"},
                "body": json.dumps({"error": "Token already used"})
                }
        
        token_data["status"] = "used"
        s3_client.put_object(
            Bucket=BUCKET_NAME,
            Key=f"{token}.json",
            Body=json.dumps(token_data),
            ContentType="application/json"
        )
        
        return {
            "statusCode": 200,
            "headers": {
                "Access-Control-Allow-Origin": "*"
            },
            "body": json.dumps({"gifter": token_data["gifter"], "giftee": token_data["giftee"]})
        }
    except ClientError as e:
        if e.response['Error']['Code'] == "NoSuchKey":
            return {"statusCode": 404, "headers": {"Access-Control-Allow-Origin": "*"},"body": json.dumps({"error": "Token not found"})}
        logger.error(f"Error retrieving token: {e}")
        return {"statusCode": 500, "headers": {"Access-Control-Allow-Origin": "*"},"body": json.dumps({"error": "Internal server error", "details": str(e)})}